dev/PipsControl/PipsControl.cpp

```diff
@@ -0,0 +1,369 @@
+}
+
+void PipsControl::OnNavigationButtonVisibilityChanged(ButtonVisibility visibility, const wstring_view collapsedStateName) {
+    if (visibility == ButtonVisibility::Collapsed) {
+        winrt::VisualStateManager::GoToState(*this, collapsedStateName, false);
+    }
+    else {
+        UpdateNavigationButtonVisualStates();
+    }
+}
+
+void PipsControl::UpdateNavigationButtonVisualStates() {
+
+    const int selectedPageIndex = SelectedPageIndex();
+    const int numberOfPages = NumberOfPages();
+    const int maxDisplayedPages = MaxDisplayedPages();
+    
+    if (!numberOfPages == 0 && maxDisplayedPages > 0) { 
+        auto const ifPreviousButtonHiddenOnEdge = selectedPageIndex == 0;
+        UpdateIndividualNavigationButtonVisualState(ifPreviousButtonHiddenOnEdge, PreviousButtonVisibility(),
+                c_previousPageButtonVisibleVisualState, c_previousPageButtonHiddenVisualState,
+                c_previousPageButtonEnabledVisualState, c_previousPageButtonDisabledVisualState);
+
+        auto const ifNextButtonHiddenOnEdge = selectedPageIndex == numberOfPages - 1;
+        UpdateIndividualNavigationButtonVisualState(ifNextButtonHiddenOnEdge, NextButtonVisibility(),
+            c_nextPageButtonVisibleVisualState, c_nextPageButtonHiddenVisualState,
+            c_nextPageButtonEnabledVisualState, c_nextPageButtonDisabledVisualState);
+    }
+}
+
+void PipsControl::OnElementPrepared(winrt::ItemsRepeater sender, winrt::ItemsRepeaterElementPreparedEventArgs args)
+{
+    if (const auto pip = args.Element().try_as<winrt::Button>()) {
+        auto const pageNumber = unbox_value<int>(pip.Tag());
+        if (unbox_value<int>(pip.Tag()) - 1 != SelectedPageIndex()) {
+            pip.Style(DefaultIndicatorStyle());
+        }
+        // Narrator says: Page 5, Button 5 of 30. Is it expected behavior?
+        winrt::AutomationProperties::SetName(pip, ResourceAccessor::GetLocalizedStringResource(SR_PipsControlPageText) + L" " + winrt::to_hstring(pageNumber));
+        winrt::AutomationProperties::SetPositionInSet(pip, pageNumber);
+        winrt::AutomationProperties::SetSizeOfSet(pip, NumberOfPages());
+        const auto buttonClickedFunc = [this](auto const& sender, auto const& args) {
+            if (const auto button = sender.try_as<winrt::Button>())
+            {
+                SelectedPageIndex(winrt::unbox_value<int>(button.Tag()) - 1);
+            }
+        };
+        pip.Click(buttonClickedFunc);
+
+    }
+}
+
+void PipsControl::ScrollToCenterOfViewport(winrt::UIElement sender)
+{
+    winrt::BringIntoViewOptions options;
+    options.VerticalAlignmentRatio(0.5);
+    options.AnimationDesired(true);
+    sender.StartBringIntoView(options);
+}
+
+
+void PipsControl::OnMaxDisplayedPagesChanged(const int oldValue) {
+    m_lastMaxDisplayedPages = oldValue;
+    auto const numberOfPages = NumberOfPages();
+    if (m_lastMaxDisplayedPages != numberOfPages) {
+        UpdateVerticalPips(numberOfPages, MaxDisplayedPages());
```

Comment:
I don't get why we don't need to update when the old value was equal to the current number of pages. If we go from 10 pages and 10 max displayed pages, to 10 pages and 5 max displayed pages we would still need to update our scrollview viewport right?